odoo/models.py

```diff
@@ -4108,17 +4096,93 @@ def _load_records(self, data_list, update=False):
                 if data.get('xml_id') and not data['xml_id'].startswith(prefix):
                     _logger.warning("Creating record %s in module %s.", data['xml_id'], module)
 
+        # we need to mark the xids of existing parents in case we're updating a
+        # db in which we previously created those xids below, otherwise
+        # _process_end will just try to delete them which would be suboptimal
+        parent_xids = {}
+        # bulk-resolve xids for each parent field
+        records = self.browse(item['record'].id for item in imd_data_list)
+        for parent_model, parent_field in self._inherits.items():
+            parent_xids[parent_field] = records.mapped(parent_field)._get_external_ids()
+
+        # create the items for each parent of each entry
+        # nb: maybe we could dedup the parent entries for "shared" parents?
+        # But needs a way to handle the noupdate which comes from the child
+        items, imd_data_list = imd_data_list, []
+        for item in items:
+            record = item['record']
+            for parent_model, parent_field in self._inherits.items():
+                if item['values'].get(parent_field):
+                    continue
+
+                parent = record[parent_field]
+                parent_xid = parent_xids[parent_field][parent.id]
+                if parent_xid:
+                    imd_data_list.append({
+                        'xml_id': parent_xid[0],
+                        'record': parent,
+                        'noupdate': item.get('noupdate', False),
+                    })
+            imd_data_list.append(item)
```

Comment:
This just keeps the crap in the database, which will never be cleaned up.
Moreover, it potentially screws up the expected deletion of record upon upgrade.
Isn't simpler to simply delete a fixed list of unexpected XML ids in pre-upgrade?

Reply:
> This just keeps the crap in the database, which will never be cleaned up.
> Moreover, it potentially screws up the expected deletion of record upon upgrade.
> Isn't simpler to simply delete a fixed list of unexpected XML ids in pre-upgrade?

I'm 100% in favor of cleaning up data rather than bloating code with tech debt, but what XML IDs are you planning to delete and why don't they have an XML ID assigned anymore?

Reply:
> This just keeps the crap in the database, which will never be cleaned up.
> Moreover, it potentially screws up the expected deletion of record upon upgrade.
> Isn't simpler to simply delete a fixed list of unexpected XML ids in pre-upgrade?

It only marks the parents of records we are updating (with a xid) as "seen", it should not keep anything more in the database: if a record is not seen during the upgrade it won't be marked and neither will its parents (unless they were marked by an other record).

Reply:
> @xmo-odoo that's it: some parent records were assigned an XML ids before, and no longer are.

They're assigned xids either way (in the old version some had multiple xids but those are not a concern for that specific issue).

The difference is that during a `-u` the parent records with *only* synthetic xids are not "traversed" anymore, so they are not added to the `loaded_xmlids` set, and `_process_end` attempts to clean them up believing they'd gotten removed.

Reply:
@odony see https://www.odoo.com/web#id=2251039&action=333&active_id=1364&model=project.task&view_type=form&cids=1&menu_id=4720 for an explanation of the issue, and the rationale behind this proposal.